### PR TITLE
Tc/schedule template

### PR DESCRIPTION
**Pull request by CedThen**

### CODE REVIEW

File: server/routes/schedules.js [hunk 1]
```diff
@@ -128,50 +132,38 @@ router.post('/schedule', certify, async (req, res, next) => {
...
-                break;
-            }
-          }),
-        );
-      }
+      await Promise.all(
+        Object.keys(changes).map(async (type) => {
+          switch (type) {
+            case 'tasks':
+              await updateTasks(changes[type], t);
```
Comment (showduhtung) [inline server/routes/schedules.js]: what is t?

Comment (showduhtung) [reply server/routes/schedules.js]: mbbe that could be described?

File: server/routes/schedules.js [hunk 2]
```diff
@@ -128,50 +132,38 @@ router.post('/schedule', certify, async (req, res, next) => {
...
-              case 'scheduleId':
-                return;
-              default:
-                break;
-            }
-          }),
-        );
-      }
+      await Promise.all(
+        Object.keys(changes).map(async (type) => {
```
Comment (showduhtung) [inline server/routes/schedules.js:154]: wonder if this could be done with object literals.

File: server/utilities/helperFunctions.js [hunk 3]
```diff
@@ -262,26 +265,32 @@ export function createColumns(start, end) {
   ];
 }
 
-export async function populateServiceData({ name, day, id }) {
+export async function populateServiceData({ name, day, id }, scheduleId) {
```
Comment (showduhtung) [inline server/utilities/helperFunctions.js]: looks a little strange but I get the point, it's annoying to move scheduleId into this. 

maybe you could just do something like 
```
export async function populateServiceData(props (or another name), scheduleId) {
  const { name, day, id } = props;
  ...
}
```

File: server/utilities/helperFunctions.js [hunk 4]
```diff
@@ -319,3 +329,105 @@ export const recurringDaysOfWeek = (start, end, dayOfWeeK) => {
   }
   return weeks;
 };
+
+export const updateEvents = async (events, t) => {
+  await Promise.all(
+    events.map(async (item, index) => {
+      const targetEvent = await db.Event.findOne({
```
Comment (showduhtung) [inline server/utilities/helperFunctions.js:337]: could destructure to be 
```
const { time, roleId } = time;
... 
return targetEvent.update( { time, roleId, order: index}, { transaction: t }); 
```

File: server/utilities/helperFunctions.js [hunk 5]
```diff
@@ -319,3 +329,105 @@ export const recurringDaysOfWeek = (start, end, dayOfWeeK) => {
...
+        where: { id: item.eventId },
+      });
+      if (targetEvent)
+        return targetEvent.update(
+          { time: item.time, roleId: item.roleId, order: index },
+          { transaction: t },
+        );
+      else {
+        const newEvent = await db.Event.create(
```
Comment (showduhtung) [inline server/utilities/helperFunctions.js:346]: heh honestly too much for my brain to handle

File: src/components/Schedule/EditServiceForm.jsx [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+    day,
+    'Must select a day of the week',
+  );
+
+  const classes = useStyles();
+
+  // console.log(`dataModel, serviceIndex`, dataModel, serviceIndex);
+
+  return (
+    <Dialog open={isOpen} onClose={onClose}>
```
Comment (showduhtung) [inline src/components/Schedule/EditServiceForm.jsx:50]: try to use the <CustomDialog /> component I created.

File: src/components/Schedule/EditServiceForm.jsx [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+      serviceName.value.length > 0 &&
+      serviceName.value.length < 32 &&
+      dayOfWeek.value >= 0
+    )
+      onSubmit(dataClone);
+    setServiceNameError(stringLengthCheck(serviceName.value));
+    setDayWeekError(dayOfWeek.value < 0);
+  }
+
+  function onDragEnd(result) {
```
Comment (showduhtung) [inline src/components/Schedule/EditServiceForm.jsx:131]: isn't this being used a lot? could this be a service used more often?

File: src/components/Schedule/EditServiceForm.jsx [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+              </MenuItem>
+              {daysOfWeek.map((d, index) => (
+                <MenuItem key={index.toString()} value={index}>
+                  {d}
+                </MenuItem>
+              ))}
+            </ValidatedSelect>
+          </div>
+          <h3>Change the order of the services</h3>
+          <DragDropList listItems={dataClone} onEnd={onDragEnd} serviceId={serviceId} />
```
Comment (showduhtung) [inline src/components/Schedule/EditServiceForm.jsx:88]: nice implementation

File: src/components/Schedule/ScheduleMain.jsx [hunk 9]
```diff
@@ -56,22 +55,17 @@ const ScheduleMain = ({
   const outerRef = useRef(null);
 
   if (!dataModel) return <div className={classes.loading}></div>;
+  // console.log(`dataModel`, dataModel);
 
   return (
     <div
       className={`main_${scheduleId}`}
-      style={{ display: isViewed ? 'block' : 'none' }}
+      style={{ visibility: isViewed ? 'visible' : 'hidden' }}
```
Comment (showduhtung) [inline src/components/Schedule/ScheduleMain.jsx]: can we rename isViewed to isVisible

**Review by showduhtung — APPROVED** [review]

File: server/routes/schedules.js (hunk 2 above)
Comment (CedThen) [reply server/routes/schedules.js:154]: yeah object literal much cleaner, idk why i spent so much time on a switch ._.

File: server/routes/schedules.js (hunk 1 above)
Comment (CedThen) [reply server/routes/schedules.js]: renamed to transaction

File: server/utilities/helperFunctions.js (hunk 3 above)
Comment (CedThen) [reply server/utilities/helperFunctions.js]: made change

File: src/components/Schedule/EditServiceForm.jsx (hunk 6 above)
Comment (CedThen) [reply src/components/Schedule/EditServiceForm.jsx:50]: mbbe in a couple PR, don't think urs was merged yet

File: src/components/Schedule/ScheduleMain.jsx (hunk 9 above)
Comment (CedThen) [reply src/components/Schedule/ScheduleMain.jsx]: k

File: src/components/Schedule/EditServiceForm.jsx (hunk 7 above)
Comment (CedThen) [reply src/components/Schedule/EditServiceForm.jsx:131]: actually this one isn't working quite right, need to figure out why first lol

**Pull request merged**